File: jandy-java-profiler/src/main/java/io/jandy/java/profiler/ThreadContext.java
package io.jandy.java.profiler;

import io.jandy.java.DataObjectBuilder;
import io.jandy.java.data.ThreadObject;
import io.jandy.java.data.TreeNode;
import io.jandy.java.util.ArrayCallStack;

import java.io.IOException;

/**
 * @author JCooky
 * @since 2016-05-23
 */
public class ThreadContext {
  private final DataObjectBuilder builder;
  private TreeNode latest, root;
  private ArrayCallStack treeNodes = new ArrayCallStack(128);
  private ThreadObject thread = new ThreadObject();

  public ThreadContext(long threadId, String threadName, DataObjectBuilder builder) throws IOException {
    this.thread.setThreadId(threadId);
    this.thread.setThreadName(threadName);
    this.builder = builder;

    this.latest = builder.getRootTreeNode();
    this.root = this.latest;
    this.thread.setRootId(this.root.getId());

    builder.save(this.latest);
  }

  public void onMethodEnter(String className, int access, String methodName, String desc) {
    treeNodes.push(latest);

    latest = this.builder.getTreeNode(className.replace('/', '.'), access, methodName, desc, latest.getId());

    latest.getAcc().setStartTime(System.nanoTime());
  }

  public void onMethodExit(Throwable throwable) {
    //    assert(latest.method.getName().equals(method.getName()));
    //    assert(latest.method.getDescriptor().equals(method.getDescriptor()));
    //    assert(latest.method.getAccess() == method.getAccess());
    //    assert(latest.method.getOwner().getName().equals(owner.getName()));
    //    assert(latest.method.getOwner().getPackageName().equals(owner.getPackageName()));

    long elapsedTime = System.nanoTime() - latest.getAcc().getStartTime();
    latest.getAcc().setElapsedTime(latest.getAcc().getElapsedTime() + elapsedTime);
    if (throwable != null)
      latest.getAcc().setException(builder.getExceptionObject(throwable));

    long time = System.nanoTime();
    builder.save(latest);
    time = System.nanoTime() - time;

    latest = treeNodes.pop();
    if (latest.getAcc() != null) {
      latest.getAcc().setElapsedTime(latest.getAcc().getElapsedTime() - time);
    }
  }

  public void onExit() {
    while (latest != root)
      onMethodExit(null);
  }

  public TreeNode getRoot() {
    return root;
  }

  public ThreadObject getThreadObject() {
    return thread;
  }
}

File: jandy-python/jandy/ThreadContext.py
import threading
import time

import math

from jandy.ThreadObject import ThreadObject


class ThreadContext:
    def __init__(self, threadId, threadName, builder):
        self.threadObject = ThreadObject()
        self.nodes = []

        self.threadObject.setThreadId(threadId)
        self.threadObject.setThreadName(threadName)
        self.context = builder

        self.latest = self.context.getRootTreeNode()
        self.root = self.latest
        self.threadObject.setRootId(self.root.get('id'))

        self.context.save(self.latest)

    def getThreadObject(self):
        return self.threadObject

    def enter(self, frame):
        if '__package__' in frame.f_globals.keys() and frame.f_globals['__package__'] == 'jandy':
            return

        # print('---- ENTER')
        n = self.context.treeNode(frame, self.latest['id'])
        # if self.context.root is None:
        #     self.context.root = n

        n['acc']['t_startTime'] = time.time()
        n['acc']['concurThreadName'] = threading.current_thread().name
        self.latest['childrenIds'].append(n['id'])

        self.nodes.append(self.latest)
        self.latest = n
        # print('ENTER - '+str(self.latest))

    def exit(self, frame, arg, excepted):
        if '__package__' in frame.f_globals.keys() and frame.f_globals['__package__'] == 'jandy':
            return

        # print('---- EXIT: '+str(frame.f_globals))
        startTime = self.latest['acc']['t_startTime']
        elapsedTime = time.time() - startTime

        self.latest['acc']['startTime'] = math.floor(startTime * 1000.0 * 1000.0 * 1000.0)
        self.latest['acc']['elapsedTime'] = math.floor(elapsedTime * 1000.0 * 1000.0 * 1000.0)
        if excepted:
            (exception, value, traceback) = arg
            self.latest['acc']['exceptionId'] = self.context.exceptionObject(exception, value, traceback)['id']

        # print('EXIT - '+str(self.latest))
        if not excepted:
            self.latest = self.nodes.pop()

        self.context.save(self.latest)